### ISSUE TITLE
Multi-component updates

### ISSUE THREAD

**Issue by athoelke** [ARM-software/psa-firmware-update-spec#12]
Devices with multiple components can require update to more than one component at the same time. Feedback in the working group demonstrated a need to support the most complex case of inter-dependent updates.

The proposed state model (see #1) is defined in relation to a single component. When multiple components are being installed, my assumption is that the API should enable this to be done with a minimum number of reboots, ideally one (please comment if this is wrong!) .

Therefore, the specification must define how the following use cases can be achieved with the API:

1. I have multiple components that require updating. There is no version dependency between them (e.g. A=v2&nbsp;&rightarrow;&nbsp;B&ge;v1, B=v2&nbsp;&rightarrow;&nbsp;A&ge;v1). I want to minimize the number of times the system reboots.
2. I have multiple components that require updating. The version dependencies between them has no cycles (e.g. A=v2&nbsp;&rightarrow;&nbsp;B&ge;v2, B=v2&nbsp;&rightarrow;&nbsp;A&ge;v1. I want to minimize the number of times the system reboots.
3. I have multiple components that require updating. The version dependencies between them is cyclic (e.g. A=v2&nbsp;&leftrightarrow;&nbsp;B=v2). I want to minimize the number of times the system reboots.

There are several ways that we can enable the Client to clearly indicate one of these scenarios to the Service. _[To be followed up later in this thread...]_

_Questions:_
* In a multi-component installation during reboot, when a component fails to install during a reboot, should this cause all components STAGED for installation fail? - or should the installer install all that do not fail, or do not depend on a component that failed to update?

**Comment by athoelke** [comment]
## Dependency checking on installation

### When a reboot is required

When updating multiple components as in the use cases above, the client might perform the update processes as follows:

```
start(A)
write(A, Av2-image)
install(A)
start(B)
write(B, Bv2-image)
install(B)
reboot()
```

**Question:** _Should an implementation be permitted to fail this sequence - and only allowed a single component to be staged at a time?_ &mdash; Propose 'no', implementations must support the above code sequence without requiring an extra reboot.

If we consider the three use cases above, for components that require a reboot:

1. No dependencies between new images: The code sequence above will work, as neither of the new components depend on anything that is not currently installed.

2. Linear dependency between new images: The behavior depends on the implementation's dependency checking:
   1.  If the implementation does no dependency checking, then the installation will succeed. If the Client has ensured that dependencies are met, then the new images will run correctly.
   2.  If the implementation defers dependency checking until the images are installed by the bootloader, then the bootloader can verify dependencies against STAGED images, where they exist for a component, rather than the currently active image. This use case will update successfully.
   3. If the implementation verifies dependencies when `install()` is called, then this code sequence will fail as Bv2 is not available when `install(A)` is called. However, given the dependencies in the new images, the Client should instead install B first, and then A - so that the dependencies of A are met when install(A) is called (Bv2 is STAGED at that point).

3. Cyclic dependency between new images: The behavior depends on the implementation's dependency checking:
   1.  If the implementation does no dependency checking, then the installation will succeed. If the Client has ensured that dependencies are met, then the new images will run correctly.
   2.  If the implementation defers dependency checking until the images are installed by the bootloader, then the bootloader can verify dependencies against STAGED images, where they exist for a component, rather than the currently active image. This use case will update successfully.
   3. If the implementation verifies dependencies when `install()` is called, then this code sequence will fail. There is no update sequence that the Client can use to resolve this. [In this scenario, the Firmware Creator could provide an additional intermediate firmware version for one of the components that avoids a cyclic dependency.]

### When no reboot is required

If the component installation does not require a reboot, then the implementation cannot defer the dependency verification after `install()` is called (which are 2.ii and 3.ii above). For this type of component, the current API can work in the following situations:

1. When there is no dependency between the new images.
2. When there is a linear dependency between the new images and the dependencies are installed first (i.e. Bv2 and then Av2).
3. When there is a cyclic dependency between the new images and the implementation does not check dependencies (this is expected to be done by the Client).

### Can we resolve scenario 3.iii with the current API?

One option is to expect a Client to interleave the component update calls, so that the implementation can see that there is a multi-component update in progress. The implementation could defer all of the dependency checking (and installation) until the final call to `install()`. The Client would then do:

```
start(A)
write(A, Av2-image)
start(B)
write(B, Bv2-image)
install(A)
install(B)
```

Although this is technically possible, it creates additional complexity:
*  The behavior of `install(A)` is different if another component is in WRITING state.
*  What is the state of component A after the call to `install(A)` that could not proceed? &mdash; is it STAGED?
*  What is the result status of the call to `install(A)`? &mdash; it cannot be `PSA_SUCCESS`. _Is this what `PSA_SUCCESS_DEPENDENCY_NEEDED` was for in v0.7?_
*  What happens to component A if the Client never calls `install(B)`, or calls `reject(B)`? &mdash; can the Client now reject update for A?
*  If there are other problems with the Av2-image, these might only be reported when `install(B)` is called. It may not be obvious that an error from the latter call requires that the new image for A is causing the problem.

### Are the other scenarios resilient to unexpected reboot?

Careful examination of the 3.ii scenario highlights another challenge with handling these use cases with the current API. If an unexpected reboot occurs after `install(A)` is called, then the bootloader/installer will attempt to install component A only, as component B has not yet been staged. This will fail, because Av2 depends on Bv2 which is not STAGED.

* Does the bootloader fail the install, and leave component A in FAILED state (with an error code)? &mdash; but from here the Client cannot attempt the install again without starting from the beginning.
* Does the bootloader detect the presence of a unstaged new image for B, and so do nothing, leaving A in STAGED state? &mdash; this changes the the expected behavior of STAGED state, requiring additional logic in the Client on retstart.

### Conclusion

Using the current API to handle the update of multiple components with cyclic dependencies results in significant additional complexity in the specification, the Client and the Service. If this use case is important for the API, then we need to consider what changes to the API would enable these use cases more effectively.

**Comment by athoelke** [comment]
## Proposed adjustment to the API for installation

To resolve the [issues identified above](https://github.com/ARM-software/psa-firmware-update-spec/issues/12#issuecomment-1182541647), we need an API that allows a Client to explicitly group a set of component updates into a synchronized installation action.

This could be done with an `install` operation that accepted multiple component Ids. To scale, this needs to accept an arbitrary number of component Ids, which is not a simple API.

Instead, this proposal is to separate the 'finished writing to the image' action from the 'install the image (or images) now' action. Then, the Client can prepare all of the images for a multi-component update, before calling `install` which acts on all images that have been prepared.

### Changes to the API

We introduce a new state, CANDIDATE, between the WRITING and STAGED/UPDATED state.
*  A component in CANDIDATE state is ready for installation, and can no longer be written to by the Client.
*  A component enters CANDIDATE state following a `finish` operation on an component in WRITING state.
*  A component in CANDIDATE state is not modified following a reboot, if the staging area is persistent.
*  A component in CANDIDATE state can be installed with the `install` operation.
*  The `install` operation acts on all components that are in CANDIDATE state atomically (i.e all are updated, or none are)
*  The `install` operation can return a status that indicates that a restart or reboot is required to continue installation.
*  It is recommended that firmware image integrity, authenticity and validity is verified when `finish` is called. This may be deferred to `install` or `reboot`, or not done at all. Dependency checking must not be done at this point.
*  It is recommended that firmware image dependencies are verified when `install` is called. This may be deferred to `reboot`, or not done at all.

The use of the `install` operation is mandatory, even for systems that only have a single component. This reduces the complexity of portable Client code which would otherwise have to conditionally call `install`, depending on system characteristics.

The Client code to update the A and B components in the original use cases would now be:

```
// Prepare new image for A
start(A)
write(A, Av2-image)
finish(A)
// Prepare new image for B
start(B)
write(B, Bv2-image)
finish(B)
// Install A and B
install()
```

**Comment by athoelke** [comment]
Here is a state model that incorporates `finish` and CANDIDATE:

![image](https://user-images.githubusercontent.com/5690199/178610883-b029ad92-3922-455c-abc7-09e83f210e48.png)

**Comment by athoelke** [comment]
### Do we need to always require atomicity of update with the proposed API?

The proposed change to the API describes `install` acting atomically on all components that are in CANDIDATE state.

It is possible to have an implementation that is capable of determining a more relaxed installation procedure based on the version dependencies between the components. For example, if the client prepares both component A and B for update, before calling `install`, but A and B have no version dependency, the implementation could safely install them independently (and perhaps fail them independently).

This capability requires the implementation to be able to determine component image versions and dependencies. This might not be possible for all implementations (for example HAL-type implementations - see #16), so the more complex behavior cannot be required.

Although this capability results in updates making better progress, in situations where some component updates fail, it comes at additional complexity within the implementation, and cannot be relied on by Clients.

### Proposal

Keep the specification simple, and require that all components installed together are installed atomically.

**Comment by athoelke** [comment]
### Calling `install` when there are already STAGED components

In a system that must reboot to complete installation, what happens with the following code sequence:

```
// Prepare new image for A
start(A)
write(A, Av2-image)
finish(A)
// Install A
install()     // returns REBOOT_REQUIRED

// Prepare new image for B
start(B)
write(B, Bv2-image)
finish(B)
// Install A and B
install()     // returns REBOOT_REQUIRED

// Restart to install
request_reboot()
```

1. Is it permitted to start preparing component B when component A is already STAGED?
2. Is it permitted to call `install()` a second time before rebooting to install A?
3. If 1 and 2 are both 'yes', then when we reboot, which of the following is best specification:

   1. The implementation MUST treat the update to A and B as independent
   2. The implementation MUST combine the updates in the two calls to `install` into a single atomic update#
   3. The implementation is permitted to EITHER combine the updates in the two calls to `install` into a single atomic update OR  treat the update to A and B as independent.
   
### Proposal

Principles:
* Allow the Client to combine updates into groups that must be installed atomically, but also update other components within the same reboot cycle.
* Allow a simpler implementation to not track multiple atomic sets of components for update.

This results in the above code sequence being valid (so (1) and (2) are 'yes'), and the specification is (3.iii) - combining updates from multiple calls to `install` is permitted, but not required.

#### Alternative

If this is considered to be too complex, then the simplifications are either to deny (2) (second call to `install` fails), or select (3.ii) to enable simpler implementations.

**Comment by athoelke** [comment]
### Calling `install` when there are components in TRIAL state

The scenario of starting a new installation while a trial is ongoing is not discussed in v0.7, though it is possible to attempt this.

There is substantial complexity in correctly handling version dependencies when the depended-on component is in TRIAL state, but could still be rejected.

I proposed that it is not permitted to call `install` if there are any components that are in TRIAL or REJECTED state. The current trial must be fully concluded before a new installation (even of other components) is attempted.

**Comment by athoelke** [comment]
Working down through the issues arising from the proposals [here](https://github.com/ARM-software/psa-firmware-update-spec/issues/12#issuecomment-1194154186) and in #24, there is an impact on Clients depending on whether the implementation maintains multiple sets of updates for atomic installation, or combines them into a single set.

The Client needs to know this in order to correctly test and `accept` or `reject` TRIAL components - must the client call `accept` for each group that it has called `install` for, or does the first call accept them all (in which case it needs to test everything before accepting anything)?

I'm not to keen on this complexity. I am inclined to recommend the alternative simplification which is to specify (3.ii) - multiple calls to `install()` prior to reboot will group all CANDIDATEs into a single set for atomic update.

**Comment by athoelke** [comment]
> I'm not to keen on this complexity. I am inclined to recommend the alternative simplification which is to specify (3.ii) - multiple calls to `install()` prior to reboot will group all CANDIDATEs into a single set for atomic update.

Further thoughts on this idea - specifying this prevents any future change to the spec that allows implementations to manage more than one set of atomic updates.

My opinion is that it would be better for v1.0 to keep the design and spec simple, and make it an error to call `install` if there are any STAGED, TRIAL or REJECTED components.

This means that there is only one 'installation process' happening at a time in the system, though this can include multiple components. We do not need to prevent component updates being downloaded (via `start`/`write`/`finish'), but newly downloaded components must wait for the current installation to complete (via `accept`) before `install` can be called again.

**Comment by athoelke** [comment]
I have rolled most of this proposal into an update of the RFC for the state model, in order to see/feel how the proposed support for simultaneous component updates (for cyclic dependencies) impacts the complexity of the API design.

The state model without this support is [here](https://github.com/ARM-software/psa-firmware-update-spec/blob/814ccb5202cad26b012f9dbbc2d7abfcd49936a3/rfc/state-model.rst).

The updated state model is [here](https://github.com/ARM-software/psa-firmware-update-spec/blob/0f2eb13c750e6a1fde6d45a5fdbca1ffee8129ad/rfc/state-model.rst).

The differences are in a single commit [here](https://github.com/ARM-software/psa-firmware-update-spec/commit/af0fd08c6f8af4b9011afa7357d03aa22c922d3c).

**Comment by athoelke** [comment]
### Decision time

A decision is required on whether the benefit of supporting atomic updates of multiple components is worth the change in API behavior proposed here.

The API changes proposed here for atomic updates result in a valuable simplification in the installation and trial behaviors relating to general component dependencies. In particular, these changes remove the requirement that the implementation can determine such dependencies, which is a better fit for systems which wish to use this API as a Platform Abstraction Layer (see #16). This makes the changes attractive, even for systems that cannot (or do not need to) support atomic updates.

The most significant loss of functionality (in comparison with the v0.7 API) is removal of the ability for multiple firmware images to be updated non-atomically in a single reboot. Some update scenarios will require additional reboots to complete.

## Proposal for v1.0

Include the API changes here into the version 1.0 API, with [this state model](https://github.com/ARM-software/psa-firmware-update-spec/blob/0f2eb13c750e6a1fde6d45a5fdbca1ffee8129ad/rfc/state-model.rst).

An implementation is permitted to support atomic multi-component updates for none, some, or all components in the system. The Firmware Image creator is assumed to know (from information from the device manufacturer/integrator), which components can be atomically updated together, and which ones cannot. The grouping of updates by the Update Client is expected to be driven by control or manifest data that is supplied via the Update Server as part of an update protocol.
